Lib/test/test_capi/test_file.py

```diff
@@ -0,0 +1,296 @@
+
+
+class TestPyFile_GetLine(_TempFileMixin, unittest.TestCase):
+    get_line = _testcapi.file_get_line
+
+    def assertGetLineNegativeIndex(self, first_line, index, eof):
+        assert index < 0
+        if first_line.endswith("\n"):
+            first_line = first_line.rstrip("\n")
+
+        with open(os_helper.TESTFN, encoding="utf-8") as f:
+            if eof:
+                self.assertRaises(EOFError, self.get_line, f, index)
+            else:
+                self.assertEqual(self.get_line(f, index), first_line)
+
+    def assertGetLine(self, first_line, eof=False):
+        with open(os_helper.TESTFN, encoding="utf-8") as f:
+            self.assertEqual(self.get_line(f, 0), first_line)
+        with open(os_helper.TESTFN, encoding="utf-8") as f:
+            self.assertEqual(self.get_line(f, 4), first_line[:4])
+        with open(os_helper.TESTFN, encoding="utf-8") as f:
+            self.assertEqual(self.get_line(f, len(first_line) + 1), first_line)
+        with open(os_helper.TESTFN, encoding="utf-8") as f:
+            self.assertEqual(self.get_line(f, _testcapi.INT_MAX), first_line)
+        self.assertGetLineNegativeIndex(first_line, -1, eof=eof)
+        self.assertGetLineNegativeIndex(first_line, _testcapi.INT_MIN, eof=eof)
+
+    def test_file_empty_line(self):
+        first_line = ""
+        with open(os_helper.TESTFN, "w", encoding="utf-8") as f:
+            f.writelines([first_line])
+        self.assertGetLine(first_line, eof=True)
+
+    def test_file_single_unicode_line(self):
+        for line_end in ("", "\n"):
+            with self.subTest(line_end=line_end):
+                first_line = "text with юникод 统一码" + line_end
+                with open(os_helper.TESTFN, "w", encoding="utf-8") as f:
+                    f.writelines([first_line])
+                self.assertGetLine(first_line)
+
+    def test_file_single_unicode_line_invalid_utf(self):
+        first_line = "\xc3\x28\n"
+        with open(os_helper.TESTFN, "w", encoding="utf-8") as f:
+            f.writelines([first_line])
+        self.assertGetLine(first_line)
+
+    def test_file_single_unicode_line_encoding_mismatch(self):
+        first_line = "text with юникод 统一码\n"
+        with open(os_helper.TESTFN, "w", encoding="utf-8") as f:
+            f.writelines([first_line])
+        with open(os_helper.TESTFN, encoding="ascii") as f:
+            self.assertRaises(UnicodeDecodeError, self.get_line, f, 0)
+
+    def test_file_multiple_unicode_lines(self):
+        first_line = "text with юникод 统一码\n"
+        with open(os_helper.TESTFN, "w", encoding="utf-8") as f:
+            f.write(first_line)
+            f.write("\n".join(["other", "line", ""]))
+        self.assertGetLine(first_line)
+
+    def test_file_get_multiple_lines(self):
+        first_line = "text with юникод 统一码\n"
+        second_line = "second line\n"
+        with open(os_helper.TESTFN, "w", encoding="utf-8") as f:
+            f.writelines([first_line, second_line])
```

Comment:
Many tests can use StringIO. E.g.
```python
f = io.StringIO('first_line\nsecond_line\n')
```

Reply:
I have an explicit test for both file object and `io.StringIO`:

```python
    def test_file_get_multiple_lines(self):
        first_line = "text with юникод 统一码\n"
        second_line = "second line\n"
        with open(os_helper.TESTFN, "w", encoding="utf-8") as f:
            f.writelines([first_line, second_line])
        with open(os_helper.TESTFN, encoding="utf-8") as f:
            self.assertEqual(self.get_line(f, 0), first_line)
            self.assertEqual(self.get_line(f, 0), second_line)

    def test_file_get_line_from_file_like(self):
        first_line = "text with юникод 统一码\n"
        second_line = "second line\n"
        contents = io.StringIO(f"{first_line}{second_line}")
        self.assertEqual(self.get_line(contents, 0), first_line)
        self.assertEqual(self.get_line(contents, 0), second_line)
```